### PR TITLE
add key filtering mechanisms to jag_conduit data reader

### PR DESCRIPTION
**Pull request by JaeseungYeom**
Add two mechanisms to filter out certain keys with the default mode, which selects all the variables. One is by the exact string matching, and the other is by prefix substring matching. Users can express such a negative selection in the prototext of the jag_conduit data reader. Especially, "jag_scalar_filters" and "jag_input_filters" rely on the former. "jag_scalar_prefix_filters" and "jag_input_prefix_filters" rely on the latter.

For example, the line below will screen out the jag scalar variables with either of the "key1" or "key2".
jag_scalar_filters: ["key1", "key2"]

Given the line below, any variable of the key that has a length no shorter than 27 and starts with the substring "image_[" is excluded.

jag_scalar_prefix_filters: [ { key_prefix: "image_["  min_len: 27} ]

### CODE REVIEW

File: src/data_readers/data_reader_jag_conduit.cpp [hunk 1]
```diff
@@ -280,7 +326,11 @@ void data_reader_jag_conduit::set_all_scalar_choices() {
   conduit::NodeConstIterator itr = n_scalar.children();
```
Comment (davidHysom) [inline src/data_readers/data_reader_jag_conduit.cpp]: Consider altering to use:
const std::vector<std::string> &names = n_scalar.shild_names();
for (auto t : names) {
  ...
}
My reasoning is that data_store can supply something similar:
const std::vector<std::string> &names = m_data_store->child_names("0/outputs/scalars");

File: src/data_readers/data_reader_jag_conduit.cpp [hunk 2]
```diff
@@ -789,7 +871,7 @@ std::vector<data_reader_jag_conduit::scalar_t> data_reader_jag_conduit::get_scal
     const conduit::Node & n_scalar = get_conduit_node(scalar_key);
```
Comment (davidHysom) [inline src/data_readers/data_reader_jag_conduit.cpp]: FYI, I think the data_store version will be:
const char* data_store_jag_conduit::get_scalar(std::string scalar_key);
  - or possibly -
std::pair<const char*, uint> data_store_jag_conduit::get_scalar(std::string scalar_key);
where 'uint' is the number of bytes.
I'm thinking that data_store will return all values as const char*, which the data_reader can then cast them as desired. Let me know if you think this won't work, or something else works better.

**Comment by JaeseungYeom** [comment]
I am making these changes. Thanks for the tip!

File: src/data_readers/data_reader_jag_conduit.cpp (hunk 2 above)
Comment (JaeseungYeom) [reply src/data_readers/data_reader_jag_conduit.cpp]: I see. It seems like there is no need for having separate get_scalar() and get_input().
We may be able to use std::pair<const char*, TypeID> instead. In which case, you need some special handling of CHAR8_STR_ID.

**Pull request merged**